arch/x86/kvm/pkvm/lapic.c

```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: GPL-2.0
+#include <linux/percpu-defs.h>
+#include <asm/apic.h>
+#include <asm/msr.h>
+#include "pkvm/lapic.h"
+
+#define LAPIC_MODE_X2APIC (X2APIC_ENABLE | XAPIC_ENABLE)
+
+struct pkvm_lapic {
+	bool ready;
+	u32 apic_id;
+};
+
+static DEFINE_PER_CPU(struct pkvm_lapic, pkvm_lapic);
+
+static int setup_lapic(struct pkvm_lapic *lapic, u64 apicbase)
+{
+	/* Lapic in x2apic mode is required by the pkvm */
+	if ((apicbase & LAPIC_MODE_X2APIC) != LAPIC_MODE_X2APIC)
```

Comment:
nit: maybe print an error message here (since now we can do that)? Then the comment will be not needed.

Reply:
Yeah, that also works. Although the error message will not be printed in non-debug mode, this still can explain why return the error code.